### PR TITLE
Introduce `@itwin/presentation-shared` package for shared code

### PR DESCRIPTION
**Pull request by grigasp**

### CODE REVIEW

File: packages/hierarchies/src/hierarchies/internal/LoggingUtils.ts
Comment (Yato333) [inline packages/hierarchies/src/hierarchies/internal/LoggingUtils.ts:1]: Can we also extract the logic for checking the log category from here?
```ts
export function log(logger: ILogger, props: LogMessageProps) {
  const severity = props.severity ?? "trace";
  if (!logger.isEnabled(props.category, severity)) {
    return;
  }
  const logFunc = getLogFunc(logger, severity);
  logFunc(props.category, props.message());
}
```

File: apps/test-app/frontend/src/components/app/App.tsx [hunk 1]
```diff
@@ -386,3 +385,8 @@ function createInstanceKeysSelectQuery(keys: InstanceKey[]): PresentationQuery {
   });
   return { query, bindings };
 }
+
+function parseFullClassName(fullClassName: string) {
```
Comment (Yato333) [inline apps/test-app/frontend/src/components/app/App.tsx:389]: Can't you use `parseFullClassName` from the shared package here?

**Review by Yato333 — COMMENTED** [review]
You can also move ResolvablePromise from test-utils to shared and reuse it in hierarchies/internal/EachValueFrom

File: packages/shared/src/test/ecsql-snippets/ECSqlJoinSnippets.test.ts [hunk 2]
```diff
@@ -5,21 +5,23 @@
...
-import { createRelationshipPathJoinClause } from "../../../hierarchies/queries/ecsql-snippets/ECSqlJoinSnippets";
-import { trimWhitespace } from "../../../hierarchies/Utils";
-import { ClassStubs, createClassStubs } from "../../Utils";
+import * as sinon from "sinon";
+import { createRelationshipPathJoinClause } from "../../shared/ecsql-snippets/ECSqlJoinSnippets";
+import * as m from "../../shared/Metadata";
+import { trimWhitespace } from "../../shared/Utils";
+import { createMetadataProviderStub } from "../MetadataProviderStub";
 
+import EC = m.EC;
```
Comment (saskliutas) [inline packages/shared/src/test/ecsql-snippets/ECSqlJoinSnippets.test.ts:14]: Why is this needed? Why not `import { EC } from "../../shared/Metadata";`?

File: packages/hierarchies/src/hierarchies/internal/LoggingUtils.ts
Comment (grigasp) [reply packages/hierarchies/src/hierarchies/internal/LoggingUtils.ts:1]: I don't see any virtue in doing that. Can you explain why you think it would help?

File: apps/test-app/frontend/src/components/app/App.tsx (hunk 1 above)
Comment (grigasp) [reply apps/test-app/frontend/src/components/app/App.tsx:389]: I could, but I'm about to remove all this selection-related code from test app, including the usage of this function, so didn't want to introduce the `@itwin/presentation-shared` dependency unnecessarily.

**Comment by grigasp** [comment]
> You can also move ResolvablePromise from test-utils to shared and reuse it in hierarchies/internal/EachValueFrom

https://github.com/iTwin/presentation/pull/522

File: packages/shared/src/test/ecsql-snippets/ECSqlJoinSnippets.test.ts (hunk 2 above)
Comment (grigasp) [reply packages/shared/src/test/ecsql-snippets/ECSqlJoinSnippets.test.ts]: Was used as a workaround for something that is now irrelevant. Cleaned up.

**Review by saskliutas — APPROVED** [review]
